Запустите, если работете в коллабе, иначе установите ручками

In [ ]:
!pip install pymorphy3
!pip install razdel
!pip install scikit-learn

In [17]:
from csv import DictReader

with open("data.csv") as dt:
    data = list(DictReader(dt))

print(*data[:5], sep="\n")

{'id': '0005998', 'score': '21.0', 'year': '2020', 'text': 'В тексте русского писателя Владимира Дудинцева говорится о человеческой жизни. Автор ставит сложную проблему соотношения продолжительности жизни и её содержания. Писатель заставляет задуматься над вопросом: важнее количество прожитых лет или то, как их прожить?\nРазмышляя о жизни, автор знакомит читателя с одной любопытной беседой, состоявшейся в научной лаборатории. Начальник рассказал своим коллегам о том, что некто якобы прожил целых 900 лет. В свою очередь, один из сотрудников поведал интересную историю про исправившегося бандита. Этот бандит понял, что всю жизни прожил зря. Он был богат, но при этом не знал настоящих ценностей: дружбы, любви, уважения окружающих. И он решил оставшиеся годы своей жизни прожить совершенно иначе — на благо обществу. Рассказчик подчеркнул, что времени у того бандита оставалось совсем мало, и человек был намерен "за год или два прожить заново свою жизнь". Этот фрагмент помогает понять, что мер

Запускать этот блок, для лемматизации

In [18]:
from pymorphy3 import MorphAnalyzer
from razdel import tokenize

morph = MorphAnalyzer()

dt = [0] * len(data)

for i in range(len(data)):
    tkn = list(tokenize(data[i]["text"]))
    dt[i] = [0] * len(tkn)
    for j in range(len(tkn)):
        dt[i][j] = morph.parse(tkn[j].text)[0].normal_form

print(*dt[:5], sep="\n")

['в', 'текст', 'русский', 'писатель', 'владимир', 'дудинцев', 'говориться', 'о', 'человеческий', 'жизнь', '.', 'автор', 'ставить', 'сложный', 'проблема', 'соотношение', 'продолжительность', 'жизнь', 'и', 'её', 'содержание', '.', 'писатель', 'заставлять', 'задуматься', 'над', 'вопрос', ':', 'важный', 'количество', 'прожить', 'год', 'или', 'то', ',', 'как', 'они', 'прожить', '?', 'размышлять', 'о', 'жизнь', ',', 'автор', 'знакомить', 'читатель', 'с', 'один', 'любопытный', 'беседа', ',', 'состояться', 'в', 'научный', 'лаборатория', '.', 'начальник', 'рассказать', 'свой', 'коллега', 'о', 'тот', ',', 'что', 'некто', 'якобы', 'прожить', 'целый', '900', 'год', '.', 'в', 'свой', 'очередь', ',', 'один', 'из', 'сотрудник', 'поведать', 'интересный', 'история', 'про', 'исправиться', 'бандит', '.', 'этот', 'бандит', 'понять', ',', 'что', 'весь', 'жизнь', 'прожить', 'зря', '.', 'он', 'быть', 'богатый', ',', 'но', 'при', 'это', 'не', 'знать', 'настоящий', 'ценность', ':', 'дружба', ',', 'любовь', ','

Запускать этот блок, для отсутствия лемматизации

In [19]:
from pymorphy3 import MorphAnalyzer
from razdel import tokenize

morph = MorphAnalyzer()

dt = [0] * len(data)

for i in range(len(data)):
    tkn = list(tokenize(data[i]["text"]))
    dt[i] = [0] * len(tkn)
    for j in range(len(tkn)):
        dt[i][j] = tkn[j].text

print(*dt[:5], sep="\n")

['В', 'тексте', 'русского', 'писателя', 'Владимира', 'Дудинцева', 'говорится', 'о', 'человеческой', 'жизни', '.', 'Автор', 'ставит', 'сложную', 'проблему', 'соотношения', 'продолжительности', 'жизни', 'и', 'её', 'содержания', '.', 'Писатель', 'заставляет', 'задуматься', 'над', 'вопросом', ':', 'важнее', 'количество', 'прожитых', 'лет', 'или', 'то', ',', 'как', 'их', 'прожить', '?', 'Размышляя', 'о', 'жизни', ',', 'автор', 'знакомит', 'читателя', 'с', 'одной', 'любопытной', 'беседой', ',', 'состоявшейся', 'в', 'научной', 'лаборатории', '.', 'Начальник', 'рассказал', 'своим', 'коллегам', 'о', 'том', ',', 'что', 'некто', 'якобы', 'прожил', 'целых', '900', 'лет', '.', 'В', 'свою', 'очередь', ',', 'один', 'из', 'сотрудников', 'поведал', 'интересную', 'историю', 'про', 'исправившегося', 'бандита', '.', 'Этот', 'бандит', 'понял', ',', 'что', 'всю', 'жизни', 'прожил', 'зря', '.', 'Он', 'был', 'богат', ',', 'но', 'при', 'этом', 'не', 'знал', 'настоящих', 'ценностей', ':', 'дружбы', ',', 'любви'

In [20]:
for i in range(len(dt)):
    dt[i] = " ".join(map(lambda x: x.lower(), dt[i]))

print(*dt[:5], sep="\n")

в тексте русского писателя владимира дудинцева говорится о человеческой жизни . автор ставит сложную проблему соотношения продолжительности жизни и её содержания . писатель заставляет задуматься над вопросом : важнее количество прожитых лет или то , как их прожить ? размышляя о жизни , автор знакомит читателя с одной любопытной беседой , состоявшейся в научной лаборатории . начальник рассказал своим коллегам о том , что некто якобы прожил целых 900 лет . в свою очередь , один из сотрудников поведал интересную историю про исправившегося бандита . этот бандит понял , что всю жизни прожил зря . он был богат , но при этом не знал настоящих ценностей : дружбы , любви , уважения окружающих . и он решил оставшиеся годы своей жизни прожить совершенно иначе — на благо обществу . рассказчик подчеркнул , что времени у того бандита оставалось совсем мало , и человек был намерен " за год или два прожить заново свою жизнь " . этот фрагмент помогает понять , что мерило человеческой жизни заключается 

**0.8** - доля датасета, уходящая на обучение модели

In [21]:
border = int(len(dt) * 0.8)
corpus = dt[:border]
corpus_ans = [float(data[i]["score"]) for i in range(border)]
test = dt[border:]
test_ans = [float(data[i]["score"]) for i in range(border, len(data))]

print(*corpus[:5], sep="\n")

в тексте русского писателя владимира дудинцева говорится о человеческой жизни . автор ставит сложную проблему соотношения продолжительности жизни и её содержания . писатель заставляет задуматься над вопросом : важнее количество прожитых лет или то , как их прожить ? размышляя о жизни , автор знакомит читателя с одной любопытной беседой , состоявшейся в научной лаборатории . начальник рассказал своим коллегам о том , что некто якобы прожил целых 900 лет . в свою очередь , один из сотрудников поведал интересную историю про исправившегося бандита . этот бандит понял , что всю жизни прожил зря . он был богат , но при этом не знал настоящих ценностей : дружбы , любви , уважения окружающих . и он решил оставшиеся годы своей жизни прожить совершенно иначе — на благо обществу . рассказчик подчеркнул , что времени у того бандита оставалось совсем мало , и человек был намерен " за год или два прожить заново свою жизнь " . этот фрагмент помогает понять , что мерило человеческой жизни заключается 

In [22]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(ngram_range=(1, 3))
train_inp = vectorizer.fit_transform(corpus)
test_inp = vectorizer.transform(test)

In [23]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

regr = LinearRegression()

regr.fit(train_inp, corpus_ans)

pred = regr.predict(test_inp)

print("Coefficients: \n", regr.coef_)
print("Mean squared error: %.2f" % mean_squared_error(test_ans, pred))
print("Coefficient of determination: %.2f" % r2_score(test_ans, pred))

print(f"Max coeff: {max(regr.coef_)}")
print(f"Min coeff: {min(regr.coef_)}")
print(f"Number of coeffs: {len(regr.coef_)}")

Coefficients: 
 [-0.09321899  0.02335146  0.02335146 ... -0.04252999 -0.04252999
 -0.04252999]
Mean squared error: 5.64
Coefficient of determination: 0.16
Max coeff: 3.4404695873067954
Min coeff: -3.6514203780006746
Number of coeffs: 1224890


# Статистика по разным способам

## Лемматизация

Coefficients: [ 0.66953742 -0.42019365 2.53753399 ... -0.19081311 -0.11508903 -0.05965664]\
Mean squared error: 7.60\
Coefficient of determination: -0.14\
Max coeff: 11.65602513382813\
Min coeff: -18.54470614733083\
Number of coeffs: 23443

### Двуграммы

Coefficients: [-0.16723512  0.0510807   0.01527822 ... -0.05688501 -0.0957871 0.03585912]\
Mean squared error: 5.59\
Coefficient of determination: 0.16\
Max coeff: 5.104032400789323\
Min coeff: -3.7083012431619347\
Number of coeffs: 374539

### Триграммы

Coefficients: [-0.13028309 0.03531143 0.03531143 ... -0.06351502 0.02660942 0.02660942]\
Mean squared error: 5.57\
Coefficient of determination: 0.17\
Max coeff: 4.350376863067781\
Min coeff: -3.419456667123719\
Number of coeffs: 1024142

## Без лемматизации

Coefficients: [ 0.1659616 -0.14171583 1.0782411 ... -0.09183369 -0.17643254 0.17386559]\
Mean squared error: 6.36\
Coefficient of determination: 0.05\
Max coeff: 7.533432605136583\
Min coeff: -6.463428064127323\
Number of coeffs: 65907

### Двуграммы

Coefficients: [-0.10491158 0.02447368 0.0128419 ... 0.02801984 -0.03560008 -0.03560008]\
Mean squared error: 5.70\
Coefficient of determination: 0.15\
Max coeff: 3.905302394711294\
Min coeff: -4.2255861916373485\
Number of coeffs: 530877

### Триграммы

Coefficients: [-0.09321899 0.02335146 0.02335146 ... -0.04252999 -0.04252999 -0.04252999]\
Mean squared error: 5.64\
Coefficient of determination: 0.16\
Max coeff: 3.4404695873067954\
Min coeff: -3.6514203780006746\
Number of coeffs: 1224890